File: aiot_ws/src/simple_ros_cpp/src/publish_map.cpp
#include "nav_msgs/msg/occupancy_grid.hpp"
#include "rclcpp/rclcpp.hpp"
#include <chrono>

using namespace std;
using namespace std::chrono_literals;

class PublishMap : public rclcpp::Node
{
public:
    explicit PublishMap()
        : Node("publish_map"), _count(0)
    {
        _pub = create_publisher<nav_msgs::msg::OccupancyGrid>("map", 10);
        _timer = create_wall_timer(1s, std::bind(&PublishMap::pub_callback, this));
    }

private:
    int _count;
    rclcpp::Publisher<nav_msgs::msg::OccupancyGrid>::SharedPtr _pub;
    rclcpp::TimerBase::SharedPtr _timer;
    void pub_callback()
    {
        auto msg = nav_msgs::msg::OccupancyGrid();
        msg.header.frame_id = "odom";
        msg.header.stamp = get_clock()->now();

        // map info
        msg.info.resolution = 0.1f;
        msg.info.width = 100;
        msg.info.height = 100;
        msg.info.origin.position.x = -(msg.info.width * msg.info.resolution) / 2;
        msg.info.origin.position.y = -(msg.info.height * msg.info.resolution) / 2;
        msg.info.origin.position.z = 0;
        msg.info.origin.orientation.x = 0;
        msg.info.origin.orientation.y = 0;
        msg.info.origin.orientation.z = 0;
        msg.info.origin.orientation.w = 1;

        msg.data.resize(msg.info.width * msg.info.height);
        for (auto &i : msg.data)
        {
            i = -1;
        }

        _pub->publish(msg);
    }
};

int main()
{
    rclcpp::init(0, nullptr);
    auto node = std::make_shared<PublishMap>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: aiot_ws/src/simple_ros/simple_ros/publish_map.py
import rclpy
from nav_msgs.msg import OccupancyGrid
from rclpy.node import Node
from std_msgs.msg import Header


class PublishMap(Node):
    def __init__(self):
        super().__init__('publish_map')
        self.pub = self.create_publisher(OccupancyGrid, 'map', 10)
        self.timer = self.create_timer(0.001, self.pub_callback)  # 1ms timer

        # Map info
        self.msg = OccupancyGrid()
        self.msg.info.resolution = 0.1
        self.msg.info.width = 100
        self.msg.info.height = 100
        self.msg.info.origin.position.x = -(self.msg.info.width * self.msg.info.resolution) / 2.0
        self.msg.info.origin.position.y = -(self.msg.info.height * self.msg.info.resolution) / 2.0
        self.msg.info.origin.position.z = 0.0
        self.msg.info.origin.orientation.x = 0.0
        self.msg.info.origin.orientation.y = 0.0
        self.msg.info.origin.orientation.z = 0.0
        self.msg.info.origin.orientation.w = 1.0

        self.msg.data = [-1 for _ in range(self.msg.info.width * self.msg.info.height)]

        self.count = 0
        self.row = 0

    def pub_callback(self):
        self.msg.header = Header()
        self.msg.header.frame_id = 'odom'
        self.msg.header.stamp = self.get_clock().now().to_msg()

        # Update map data
        index = self.count + (self.msg.info.width * self.row)
        if self.msg.data[index] == -1:
            self.msg.data[index] = 100
        else:
            self.msg.data[index] = -1

        self.count += 1
        if self.count >= self.msg.info.width:
            self.count = 0
            self.row += 1

        if self.row >= self.msg.info.height:
            self.row = 0

        # Publish the map
        self.pub.publish(self.msg)

def main(args=None):
    rclpy.init(args=args)
    node = PublishMap()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()

if __name__ == '__main__':
    main()
